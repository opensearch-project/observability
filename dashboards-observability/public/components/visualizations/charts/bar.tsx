/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';

import { take, merge } from 'lodash';
import { Plt } from '../plotly/plot';
import { LONG_CHART_COLOR, PLOTLY_COLOR } from '../../../../common/constants/shared';

export const Bar = ({
  visualizations,
  barConfig = {},
  layoutConfig = {},
  isUniColor = false,
}: any) => {
  const {
    data,
    metadata: { fields },
  } = visualizations;
  const stackLength = fields.length - 1;

  // Individual bars have different colors
  // when: stackLength = 1 and length of result buckets < 16 and chart is not unicolor 
  // Else each stacked bar has its own color using colorway
  let marker = {};
  if (stackLength == 1 && data[fields[stackLength].name].length < 16 && !isUniColor) {
    marker = {
      color: data[fields[stackLength].name].map((_: string, index: number) => {
        return PLOTLY_COLOR[index % PLOTLY_COLOR.length];
      }),
    };
  }

  const barValues = take(fields, stackLength > 0 ? stackLength : 1).map((field: any) => {
    return {
      x: barConfig.orientation !== 'h' ? data[fields[stackLength].name] : data[field.name],
      y: barConfig.orientation !== 'h' ? data[field.name] : data[fields[stackLength].name],
      type: 'bar',
      marker: marker,
      name: field.name,
      ...barConfig,
    };
  });

  const barLayoutConfig = merge(
    {
      xaxis: {
        automargin: true,
      },
      yaxis: {
        automargin: true,
      },
    },
    layoutConfig
  );

  // If chart has length of result buckets < 16
  // then use the LONG_CHART_COLOR for all the bars in the chart
  const plotlyColorway =
    data[fields[stackLength].name].length < 16 ? PLOTLY_COLOR : [LONG_CHART_COLOR];

  return (
    <Plt
      data={barValues}
      layout={{
        colorway: plotlyColorway,
        xaxis: {
          showgrid: false,
          visible: true,
        },
        yaxis: {
          showgrid: false,
          visible: true,
        },
        ...barLayoutConfig,
      }}
      config={barConfig}
    />
  );
};
