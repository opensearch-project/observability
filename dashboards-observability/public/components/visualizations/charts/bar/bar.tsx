/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { isEmpty, last, take } from 'lodash';
import { Plt } from '../../plotly/plot';
import { LONG_CHART_COLOR, PLOTLY_COLOR } from '../../../../../common/constants/shared';
import { AvailabilityUnitType } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_availability';
import { ThresholdUnitType } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_thresholds';

export const Bar = ({ visualizations, layout, config }: any) => {
  const { vis } = visualizations;
  const {
    data,
    metadata: { fields },
  } = visualizations.data.rawVizData;
  const { isUniColor } = vis.visConfig;
  const lastIndex = fields.length - 1;
  const {
    dataConfig = {},
    layoutConfig = {},
    availabilityConfig = {},
  } = visualizations?.data?.userConfigs;
  const xaxis =
    dataConfig.valueOptions && dataConfig.valueOptions.xaxis ? dataConfig.valueOptions.xaxis : [];
  const yaxis =
    dataConfig.valueOptions && dataConfig.valueOptions.xaxis ? dataConfig?.valueOptions.yaxis : [];
  const barOrientation =
    dataConfig?.chartOptions?.orientation &&
    dataConfig.chartOptions.orientation[0] &&
    dataConfig.chartOptions.orientation[0].orientationId
      ? dataConfig.chartOptions.orientation[0].orientationId
      : visualizations.vis.orientation;
  const { defaultAxes } = visualizations.data;

  const isVertical = barOrientation === 'v';

  // Individual bars have different colors
  // when: stackLength = 1 and length of result buckets < 16 and chart is not unicolor
  // Else each stacked bar has its own color using colorway
  let marker = {};
  if (lastIndex === 1 && data[fields[lastIndex].name].length < 16 && !isUniColor) {
    marker = {
      color: data[fields[lastIndex].name].map((_: string, index: number) => {
        return PLOTLY_COLOR[index % PLOTLY_COLOR.length];
      }),
    };
  }

  let valueSeries;
  if (!isEmpty(xaxis) && !isEmpty(yaxis)) {
    valueSeries = isVertical ? [...yaxis] : [...xaxis];
  } else {
    valueSeries = defaultAxes.yaxis || take(fields, lastIndex > 0 ? lastIndex : 1);
  }

  // determine category axis
  let bars = valueSeries.map((field: any) => {
    return {
      x: isVertical
        ? data[!isEmpty(xaxis) ? xaxis[0].label : fields[lastIndex].name]
        : data[field.name],
      y: isVertical
        ? data[field.name]
        : data[!isEmpty(yaxis) ? yaxis[0]?.label : fields[lastIndex].name],
      type: vis.type,
      marker,
      name: field.name,
      orientation: barOrientation,
    };
  });

  // If chart has length of result buckets < 16
  // then use the LONG_CHART_COLOR for all the bars in the chart
  const plotlyColorway =
    data[fields[lastIndex].name].length < 16 ? PLOTLY_COLOR : [LONG_CHART_COLOR];

  const mergedLayout = {
    colorway: plotlyColorway,
    ...layout,
    ...(layoutConfig.layout && layoutConfig.layout),
    title: dataConfig?.panelOptions?.title || layoutConfig.layout?.title || '',
    barmode:
      dataConfig?.chartOptions?.mode &&
      dataConfig.chartOptions.mode[0] &&
      dataConfig.chartOptions.mode[0].modeId
        ? dataConfig.chartOptions.mode[0].modeId
        : '',
  };

  if (dataConfig.thresholds || availabilityConfig.level) {
    const thresholdTraces = {
      x: [],
      y: [],
      mode: 'text',
      text: [],
    };
    const thresholds = dataConfig.thresholds ? dataConfig.thresholds : [];
    const levels = availabilityConfig.level ? availabilityConfig.level : [];

    const mapToLine = (list: ThresholdUnitType[] | AvailabilityUnitType[], lineStyle: any) => {
      return list.map((thr: ThresholdUnitType) => {
        thresholdTraces.x.push(
          data[!isEmpty(xaxis) ? xaxis[xaxis.length - 1]?.label : fields[lastIndex].name][0]
        );
        thresholdTraces.y.push(thr.value * (1 + 0.06));
        thresholdTraces.text.push(thr.name);
        return {
          type: 'line',
          x0: data[!isEmpty(xaxis) ? xaxis[0]?.label : fields[lastIndex].name][0],
          y0: thr.value,
          x1: last(data[!isEmpty(xaxis) ? xaxis[0]?.label : fields[lastIndex].name]),
          y1: thr.value,
          name: thr.name || '',
          opacity: 0.7,
          line: {
            color: thr.color,
            width: 3,
            ...lineStyle,
          },
        };
      });
    };

    mergedLayout.shapes = [...mapToLine(thresholds, { dash: 'dashdot' }), ...mapToLine(levels, {})];
    bars = [...bars, thresholdTraces];
  }

  const mergedConfigs = {
    ...config,
    ...(layoutConfig.config && layoutConfig.config),
  };

  return <Plt data={bars} layout={mergedLayout} config={mergedConfigs} />;
};
