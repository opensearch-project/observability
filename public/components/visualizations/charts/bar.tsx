/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { Plt } from '../plotly/plot';

export const Bar = ({
  xvalues,
  yvalues,
  name,
  layoutConfig,
}: any) => {
  return (
    <Plt 
      data={[
        {
          marker: {
            color: '#006BB4'
          },
          x: xvalues,
          y: yvalues,
          type: 'bar',
          name,
        }
      ]}
      layout={{ 
        xaxis: {
          fixedrange: true,
          showgrid: false,
          visible: true
        },
        yaxis: {
          fixedrange: true,
          showgrid: false,
          visible: true
        },
        ...layoutConfig
      }}
    />  
  );
};