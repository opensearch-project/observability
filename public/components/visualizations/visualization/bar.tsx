/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { Plt } from '../plotly/plot_template';

export const Bar = (props: any) => {
  return (
    <Plt 
      data={[
        {
          marker: {
            color: '#006BB4'
          },
          x: ['13:00:00', '13:00:30', '13:01:00', '13:01:30', '13:02:00', '13:02:30','13:03:00', '13:03:30', '13:04:00', '13:04:30', '13:05:00', '13:05:30', '13:06:00', '13:06:30', '13:07:00'],
          y: [12, 2, 7, 6, 0, 0, 8, 28, 47, 33, 13, 10, 11, 27, 32],
          type: 'bar',
          name: 'Count Distribution'
        }
      ]}
      layout={{ 
        width: '100%', 
        height: '100%',
        xaxis: {
          fixedrange: true,
          showgrid: false,
          visible: true
        },
        yaxis: {
          fixedrange: true,
          showgrid: false,
          visible: true
        },
      }}
    />  
  );
};