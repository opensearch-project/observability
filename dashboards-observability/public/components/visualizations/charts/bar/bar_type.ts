/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Bar } from './bar';
import { getPlotlySharedConfigs, getPlotlyCategory } from '../shared/shared_configs';
import { LensIconChartBar } from '../../assets/chart_bar';
import { VizDataPanel } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/default_vis_editor';
import { ConfigEditor } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/json_editor';
import {
  ConfigLegend,
  InputFieldItem,
  ConfigColorTheme,
  SliderConfig,
  ConfigBarChartStyles,
  ButtonGroupItem,
  ConfigAvailability,
} from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls';
import { BarOrientation, DEFAULT_CHART_STYLES } from '../../../../../common/constants/shared';
import { fetchConfigObject } from '../../../../components/event_analytics/utils/utils';
import { VIS_CHART_TYPES } from '../../../../../common/constants/shared';
import { DEFAULT_BAR_CHART_STYLES } from '../../../../../common/constants/explorer';

const sharedConfigs = getPlotlySharedConfigs();
const VIS_CATEGORY = getPlotlyCategory();

const { LegendPosition, ShowLegend, LabelAngle, FillOpacity, LineWidth } = DEFAULT_CHART_STYLES;
const { BarMode, GroupWidth, BarWidth } = DEFAULT_BAR_CHART_STYLES;
const isHorizontalBar = (paramstype: string) =>
  paramstype === VIS_CHART_TYPES.HorizontalBar ? true : false;

export const createBarTypeDefinition = (params: any) => ({
  name: params.type || 'bar',
  type: 'bar',
  id: params.type || 'bar',
  label: isHorizontalBar(params.type) ? 'Horizontal bar' : 'Vertical bar',
  fulllabel: isHorizontalBar(params.type) ? 'Horizontal bar' : 'Vertical bar',
  icontype: isHorizontalBar(params.type) ? 'visBarHorizontalStacked' : 'visBarVerticalStacked',
  selection: {
    dataLoss: 'nothing',
  },
  category: VIS_CATEGORY.BASICS,
  icon: LensIconChartBar,
  categoryaxis: 'xaxis',
  seriesaxis: 'yaxis',
  orientation: isHorizontalBar(params.type) ? BarOrientation.horizontal : BarOrientation.vertical,
  mode: BarMode,
  labelangle: LabelAngle,
  linewidth: LineWidth,
  fillopacity: FillOpacity,
  groupwidth: GroupWidth,
  barwidth: BarWidth,
  showlegend: ShowLegend,
  legendposition: LegendPosition,
  component: Bar,
  editorconfig: {
    panelTabs: [
      {
        id: 'data-panel',
        name: 'Style',
        mapTo: 'dataConfig',
        editor: VizDataPanel,
        sections: [
          fetchConfigObject('Tooltip', {
            options: [
              { name: 'All', id: 'all' },
              { name: 'Dimension', id: 'x' },
              { name: 'Series', id: 'y' },
            ],
            defaultSelections: [{ name: 'All', id: 'all' }],
          }),
          {
            id: 'legend',
            name: 'Legend',
            editor: ConfigLegend,
            mapTo: 'legend',
            schemas: [
              {
                name: 'Show legend',
                mapTo: 'showLegend',
                component: null,
                props: {
                  options: [
                    { name: 'Show', id: 'show' },
                    { name: 'Hidden', id: 'hidden' },
                  ],
                  defaultSelections: [{ name: 'Show', id: ShowLegend }],
                },
              },
              {
                name: 'Position',
                mapTo: 'position',
                component: null,
                props: {
                  options: [
                    { name: 'Right', id: 'v' },
                    { name: 'Bottom', id: 'h' },
                  ],
                  defaultSelections: [{ name: 'Right', id: LegendPosition }],
                },
              },
              {
                title: 'Legend size',
                name: 'Legend size',
                component: InputFieldItem,
                mapTo: 'legendSize',
                eleType: 'input',
              },
            ],
          },
          {
            id: 'chart_styles',
            name: 'Chart styles',
            editor: ConfigBarChartStyles,
            mapTo: 'chartStyles',
            schemas: [
              {
                name: 'Mode',
                component: ButtonGroupItem,
                mapTo: 'mode',
                eleType: 'buttons',
                props: {
                  options: [
                    { name: 'Group', id: BarMode },
                    { name: 'Stack', id: 'stack' },
                  ],
                  defaultSelections: [{ name: 'Group', id: BarMode }],
                },
              },
              {
                name: 'Label size',
                component: InputFieldItem,
                mapTo: 'labelSize',
                eleType: 'input',
              },
              {
                name: 'Rotate bar labels',
                component: SliderConfig,
                mapTo: 'rotateBarLabels',
                eleType: 'slider',
                defaultState: LabelAngle,
                props: {
                  ticks: [
                    { label: '-90°', value: -90 },
                    { label: '-45°', value: -45 },
                    { label: '0°', value: 0 },
                    { label: '45°', value: 45 },
                    { label: '90°', value: 90 },
                  ],
                  showTicks: true,
                  min: -90,
                  max: 90,
                },
              },
              {
                name: 'Group width',
                component: SliderConfig,
                mapTo: 'groupWidth',
                defaultState: GroupWidth,
                props: {
                  max: 1,
                  step: 0.01,
                },
                eleType: 'slider',
              },
              {
                name: 'Bar width',
                component: SliderConfig,
                mapTo: 'barWidth',
                defaultState: BarWidth,
                props: {
                  max: 1,
                  step: 0.01,
                },
                eleType: 'slider',
              },
              {
                name: 'Line width',
                component: SliderConfig,
                mapTo: 'lineWidth',
                defaultState: LineWidth,
                props: {
                  max: 10,
                },
                eleType: 'slider',
              },
              {
                name: 'Fill opacity',
                component: SliderConfig,
                mapTo: 'fillOpacity',
                defaultState: FillOpacity,
                props: {
                  max: 100,
                },
                eleType: 'slider',
              },
            ],
          },
        ],
      },
      {
        id: 'availability-panel',
        name: 'Availability',
        mapTo: 'availabilityConfig',
        editor: ConfigAvailability,
      },
    ],
  },
  visconfig: {
    layout: {
      ...sharedConfigs.layout,
    },
    config: {
      ...sharedConfigs.config,
    },
    isUniColor: false,
  },
});
