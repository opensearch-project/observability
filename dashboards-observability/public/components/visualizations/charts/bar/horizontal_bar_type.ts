import { Bar } from './bar';
import { getPlotlySharedConfigs, getPlotlyCategory } from '../shared/shared_configs';
import { LensIconChartBar } from '../../assets/chart_bar';
import { VizDataPanel } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/default_vis_editor';
import { ConfigEditor } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/json_editor';
import { ConfigValueOptions } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls';

const sharedConfigs = getPlotlySharedConfigs();
const VIS_CATEGORY = getPlotlyCategory();

export interface BarTypeParams {}

export const createHorizontalBarTypeDefinition = (params: BarTypeParams = {}) => ({
  id: 'horizontal_bar',
  name: 'horizontal_bar',
  type: 'bar',
  label: 'Bar',
  fulllabel: 'Bar',
  icontype: 'visBarVerticalStacked',
  selection: {
    dataLoss: 'nothing',
  },
  category: VIS_CATEGORY.BASICS,
  icon: LensIconChartBar,
  categoryaxis: 'xaxis',
  seriesaxis: 'yaxis',
  orientation: 'h',
  component: Bar,
  editorconfig: {
    panelTabs: [
      {
        id: 'data-panel',
        name: 'Style',
        mapTo: 'dataConfig',
        editor: VizDataPanel,
        sections: [
          {
            id: 'value_options',
            name: 'Value options',
            editor: ConfigValueOptions,
            mapTo: 'valueOptions',
            schemas: [
              {
                name: 'X-axis',
                isSingleSelection: false,
                component: null,
                mapTo: 'xaxis',
              },
              {
                name: 'Y-axis',
                isSingleSelection: false,
                component: null,
                mapTo: 'yaxis',
              },
            ],
          },
          {
            id: 'chart_options',
            name: 'Chart options',
            editor: ConfigValueOptions,
            mapTo: 'chartOptions',
            schemas: [
              {
                name: 'Orientation',
                isSingleSelection: true,
                component: null,
                mapTo: 'orientation',
                props: {
                  dropdownList: [
                    { name: 'Vertical', orientationId: 'v' },
                    { name: 'Horizontal', orientationId: 'h' },
                  ],
                  defaultSelections: [{ name: 'Horizontal', orientationId: 'h' }],
                },
              },
              {
                name: 'Mode',
                isSingleSelection: true,
                component: null,
                mapTo: 'mode',
                props: {
                  dropdownList: [
                    { name: 'Group', modeId: 'group' },
                    { name: 'Stack', modeId: 'stack' },
                  ],
                  defaultSelections: [{ name: 'Group', modeId: 'group' }],
                },
              },
            ],
          },
        ],
      },
      {
        id: 'style-panel',
        name: 'Layout',
        mapTo: 'layoutConfig',
        editor: ConfigEditor,
        content: [],
      },
    ],
  },
  visconfig: {
    layout: {
      ...sharedConfigs.layout,
    },
    config: {
      ...sharedConfigs.config,
    },
    isUniColor: false,
  },
});
