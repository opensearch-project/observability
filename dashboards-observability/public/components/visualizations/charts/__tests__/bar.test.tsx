/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import { Bar } from '../bar/bar';
import {
  LAYOUT_CONFIG,
  TEST_VISUALIZATIONS_DATA
} from '../../../../../test/event_analytics_constants';

describe('Veritcal Bar component', () => {
  configure({ adapter: new Adapter() });

  it('Renders veritcal bar component', async () => {
    const wrapper = mount(
      <Bar
        visualizations={TEST_VISUALIZATIONS_DATA}
        layout={LAYOUT_CONFIG}
        config={true}
      />
    );

    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
