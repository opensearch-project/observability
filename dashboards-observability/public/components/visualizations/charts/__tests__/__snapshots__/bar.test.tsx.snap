// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar component Renders bar component 1`] = `
<Bar
  config={true}
  layout={
    Object {
      "colorway": Array [
        "#8C55A3",
      ],
      "height": 220,
      "margin": Object {
        "b": 15,
        "l": 60,
        "pad": 0,
        "r": 10,
        "t": 30,
      },
      "showlegend": true,
    }
  }
  visualizations={
    Object {
      "data": Object {
        "appData": Object {
          "fromApp": false,
        },
        "defaultAxes": Object {},
        "indexFields": Object {
          "availableFields": Array [
            Object {
              "name": "agent",
              "type": "string",
            },
            Object {
              "name": "bytes",
              "type": "long",
            },
            Object {
              "name": "clientip",
              "type": "ip",
            },
            Object {
              "name": "event",
              "type": "struct",
            },
            Object {
              "name": "extension",
              "type": "string",
            },
            Object {
              "name": "geo",
              "type": "struct",
            },
            Object {
              "name": "host",
              "type": "string",
            },
            Object {
              "name": "index",
              "type": "string",
            },
            Object {
              "name": "ip",
              "type": "ip",
            },
            Object {
              "name": "machine",
              "type": "struct",
            },
            Object {
              "name": "memory",
              "type": "double",
            },
            Object {
              "name": "message",
              "type": "string",
            },
            Object {
              "name": "phpmemory",
              "type": "long",
            },
            Object {
              "name": "referer",
              "type": "string",
            },
            Object {
              "name": "request",
              "type": "string",
            },
            Object {
              "name": "response",
              "type": "string",
            },
            Object {
              "name": "tags",
              "type": "string",
            },
            Object {
              "name": "timestamp",
              "type": "timestamp",
            },
            Object {
              "name": "url",
              "type": "string",
            },
            Object {
              "name": "utc_time",
              "type": "timestamp",
            },
          ],
          "queriedFields": Array [
            Object {
              "name": "double_per_ip_bytes",
              "type": "long",
            },
            Object {
              "name": "host",
              "type": "text",
            },
            Object {
              "name": "ip_count",
              "type": "integer",
            },
            Object {
              "name": "per_ip_bytes",
              "type": "long",
            },
            Object {
              "name": "resp_code",
              "type": "text",
            },
            Object {
              "name": "sum_bytes",
              "type": "long",
            },
          ],
          "selectedFields": Array [],
          "unselectedFields": Array [],
        },
        "query": Object {},
        "rawVizData": Object {
          "data": Object {
            "count()": Array [
              154,
              1753,
              116,
              468,
              1964,
              219,
            ],
            "tags": Array [
              "error",
              "info",
              "login",
              "security",
              "success",
              "warning",
            ],
          },
          "jsonData": Array [
            Object {
              "count()": 154,
              "tags": "error",
            },
            Object {
              "count()": 1753,
              "tags": "info",
            },
            Object {
              "count()": 116,
              "tags": "login",
            },
            Object {
              "count()": 468,
              "tags": "security",
            },
            Object {
              "count()": 1964,
              "tags": "success",
            },
            Object {
              "count()": 219,
              "tags": "warning",
            },
          ],
          "metadata": Object {
            "fields": Array [
              Object {
                "name": "count()",
                "type": "integer",
              },
              Object {
                "name": "tags",
                "type": "text",
              },
            ],
            "size": 6,
            "status": 200,
          },
        },
        "userConfigs": Object {},
      },
      "vis": Object {
        "category": "Visualizations",
        "categoryAxis": "xaxis",
        "component": [Function],
        "editorConfig": Object {
          "panelTabs": Array [
            Object {
              "editor": [Function],
              "id": "data-panel",
              "mapTo": "dataConfig",
              "name": "Data",
              "sections": Array [
                Object {
                  "editor": [Function],
                  "id": "value_options",
                  "mapTo": "valueOptions",
                  "name": "Value options",
                  "schemas": Array [
                    Object {
                      "component": null,
                      "isSingleSelection": false,
                      "mapTo": "xaxis",
                      "name": "X-axis",
                    },
                    Object {
                      "component": null,
                      "isSingleSelection": false,
                      "mapTo": "yaxis",
                      "name": "Y-axis",
                    },
                  ],
                },
                Object {
                  "editor": [Function],
                  "id": "chart_options",
                  "mapTo": "chartOptions",
                  "name": "Chart options",
                  "schemas": Array [
                    Object {
                      "component": null,
                      "isSingleSelection": true,
                      "mapTo": "orientation",
                      "name": "Orientation",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "name": "Vertical",
                            "orientationId": "v",
                          },
                        ],
                        "dropdownList": Array [
                          Object {
                            "name": "Vertical",
                            "orientationId": "v",
                          },
                          Object {
                            "name": "Horizontal",
                            "orientationId": "h",
                          },
                        ],
                      },
                    },
                    Object {
                      "component": null,
                      "isSingleSelection": true,
                      "mapTo": "mode",
                      "name": "Mode",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "modeId": "group",
                            "name": "Group",
                          },
                        ],
                        "dropdownList": Array [
                          Object {
                            "modeId": "group",
                            "name": "Group",
                          },
                          Object {
                            "modeId": "stack",
                            "name": "Stack",
                          },
                        ],
                      },
                    },
                  ],
                },
              ],
            },
            Object {
              "content": Array [],
              "editor": [Function],
              "id": "style-panel",
              "mapTo": "layoutConfig",
              "name": "Layout",
            },
            Object {
              "editor": [Function],
              "id": "availability-panel",
              "mapTo": "availabilityConfig",
              "name": "Availability",
            },
          ],
        },
        "fullLabel": "Bar",
        "icon": [Function],
        "iconType": "visBarVerticalStacked",
        "id": "bar",
        "label": "Bar",
        "name": "bar",
        "orientation": "v",
        "selection": Object {
          "dataLoss": "nothing",
        },
        "seriesAxis": "yaxis",
        "type": "bar",
        "visConfig": Object {
          "config": Object {
            "displaylogo": false,
            "responsive": true,
          },
          "isUniColor": false,
          "layout": Object {
            "height": 500,
            "legend": Object {
              "orientation": "v",
              "traceorder": "normal",
            },
            "margin": Object {
              "b": 30,
              "l": 60,
              "pad": 0,
              "r": 30,
              "t": 50,
            },
            "showlegend": true,
          },
        },
      },
    }
  }
>
  <Plt
    config={Object {}}
    data={
      Array [
        Object {
          "marker": Object {
            "color": Array [
              "#3CA1C7",
              "#8C55A3",
              "#DB748A",
              "#F2BE4B",
              "#68CCC2",
              "#2A7866",
            ],
          },
          "name": "count()",
          "orientation": "v",
          "type": "bar",
          "x": Array [
            "error",
            "info",
            "login",
            "security",
            "success",
            "warning",
          ],
          "y": Array [
            154,
            1753,
            116,
            468,
            1964,
            219,
          ],
        },
      ]
    }
    layout={
      Object {
        "barmode": "",
        "colorway": Array [
          "#8C55A3",
        ],
        "height": 220,
        "margin": Object {
          "b": 15,
          "l": 60,
          "pad": 0,
          "r": 10,
          "t": 30,
        },
        "showlegend": true,
        "title": "",
      }
    }
  >
    <PlotlyComponent
      config={
        Object {
          "displayModeBar": false,
        }
      }
      data={
        Array [
          Object {
            "marker": Object {
              "color": Array [
                "#3CA1C7",
                "#8C55A3",
                "#DB748A",
                "#F2BE4B",
                "#68CCC2",
                "#2A7866",
              ],
            },
            "name": "count()",
            "orientation": "v",
            "type": "bar",
            "x": Array [
              "error",
              "info",
              "login",
              "security",
              "success",
              "warning",
            ],
            "y": Array [
              154,
              1753,
              116,
              468,
              1964,
              219,
            ],
          },
        ]
      }
      debug={false}
      divId="explorerPlotComponent"
      layout={
        Object {
          "autosize": true,
          "barmode": "",
          "colorway": Array [
            "#8C55A3",
          ],
          "height": 220,
          "hovermode": "closest",
          "legend": Object {
            "orientation": "h",
            "traceorder": "normal",
          },
          "margin": Object {
            "b": 15,
            "l": 60,
            "pad": 0,
            "r": 10,
            "t": 30,
          },
          "showlegend": true,
          "title": "",
          "xaxis": Object {
            "automargin": true,
            "rangemode": "normal",
            "showgrid": true,
            "zeroline": false,
          },
          "yaxis": Object {
            "rangemode": "normal",
            "showgrid": true,
            "zeroline": false,
          },
        }
      }
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
      useResizeHandler={true}
    >
      <div
        id="explorerPlotComponent"
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </PlotlyComponent>
  </Plt>
</Bar>
`;
